labelbox/schema/data_row.py

```diff
@@ -17,6 +21,79 @@
 logger = logging.getLogger(__name__)
 
 
+class KeyType(str, Enum):
+    ID = 'ID'
+    """An existing CUID"""
+    GKEY = 'GKEY'
+    """A Global key, could be existing or non-existing"""
+    AUTO = 'AUTO'
+    """The key will be auto-generated. Only usable for creates"""
+
+
+class DataRowKey(BaseModel, ABC):
+    """
+    The DataRowKey class is a unique ID abstraction that allows us to reference
+    a DataRow by either a Global key or CUID
+    """
+    type: KeyType
+    value: str
+
+
+class DataRowIdKey(DataRowKey):
+    """
+    This represents a data row identifier key that is provided by Labelbox upon data row creation.
+    """
+
+    def __init__(self, value: str):
+        super().__init__(type=KeyType.ID, value=value)
+
+
+class DataRowGlobalKey(DataRowKey):
+    """
+    This represents a unique data row key that is provided by the user.
+    """
+
+    def __init__(self, value: str):
+        super().__init__(type=KeyType.GKEY, value=value)
+
+
+class DataRowAutoKey(DataRowKey):
+    """
+    This represents a key for a create-only data row.
+    """
+
+    def __init__(self):
+        super().__init__(type=KeyType.AUTO, value="")
+
+
+class DataRowMetadataSpec(BaseModel):
+    schema_id: Optional[str]
+    value: Any
+    name: Optional[str]
+
+
+class DataRowAttachmentSpec(BaseModel):
+    type: AttachmentType
+    value: str
+    name: Optional[str]
+
+
+class DataRowSpec(BaseModel):
```

Comment:
I'm not sure we should include these models here - they are only used for `upsert` and have very generic names which might lead to confusion on when to use and why and IMO we do not want to set any new standards as part of this change.

We can think on this more but we most likely want to get buy-in from code-owners and/or MLSE.

Reply:
in the follow up refactor ticket, I was gonna reuse this for the create methods
IMO we better start using types to make this library easier to SDK. Without them clients need to either read the docs or the underlying code itself, especially when we use `*args` and `**kwargs` making it even more complicated to understand what variables are supported

Reply:
Can you provide more details re: what you plan to refactor?

I'm worries the scope is too large - you mention refactoring `create_data_rows` but you also need to be mindful that we do not break existing implementations as that would seemingly imply a major upgrade to the SDK.

Personally, I believe we should aim our sights lower with a minimal integration, minimizing changes to existing patterns.

Per above, if you want to continue down this path we likely need to get buy-in from the SDK team as they would likley own this work.

Reply:
I'd love to have types in all these methods, but I think Matt's right; it's not trivial to add typing to these methods without breaking backward compatibility, and until we do that, we'd have 2 similar methods taking in arguments in different formats.

Since we can't add typing to other methods immediately, we should aim for consistency for now. If we do upgrade to typed arguments, we can do it in all the relevant methods at once, in a single PR, and announce it in patch notes for the relevant SDK version.

Could you file a ticket (if we don't have one already) for adding types here and put it in backlog so this doesn't get lost?